tracing-subscriber/src/fmt/fmt_layer.rs

```diff
@@ -0,0 +1,457 @@
+///
+/// [extensions]: ../registry/extensions/index.html
+#[derive(Default)]
+pub struct FormattedFields<E> {
+    _format_event: PhantomData<fn(E)>,
+    /// The formatted fields of a span.
+    pub fmt_fields: String,
+}
+
+impl<E> fmt::Debug for FormattedFields<E> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("FormattedFields")
+            .field("fmt_fields", &self.fmt_fields)
+            .finish()
+    }
+}
+
+// === impl FmtLayer ===
+
+impl<S, N, E, W> Layer<S> for FmtLayer<S, N, E, W>
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
+    N: for<'writer> FormatFields<'writer> + 'static,
+    E: FormatEvent<S, N> + 'static,
+    W: MakeWriter + 'static,
+{
+    fn new_span(&self, attrs: &Attributes<'_>, id: &Id, ctx: Context<'_, S>) {
+        let span = ctx.span(id).expect("Span not found, this is a bug");
+        let mut extensions = span.extensions_mut();
+
+        let mut buf = String::new();
+        if self.fmt_fields.format_fields(&mut buf, attrs).is_ok() {
+            let fmt_fields = FormattedFields {
+                fmt_fields: buf,
+                _format_event: PhantomData::<fn(N)>,
+            };
+            extensions.insert(fmt_fields);
+        }
+    }
+
+    fn on_record(&self, id: &Id, values: &Record<'_>, ctx: Context<'_, S>) {
+        let span = ctx.span(id).expect("Span not found, this is a bug");
+        let mut extensions = span.extensions_mut();
+
+        let mut buf = String::new();
+        if self.fmt_fields.format_fields(&mut buf, values).is_ok() {
+            let buf = match extensions.get_mut::<FormattedFields<Self>>() {
+                Some(fields) => format!("{}{}", fields.fmt_fields, buf),
+                None => buf,
+            };
+            let fmt_fields = FormattedFields {
+                fmt_fields: buf,
+                _format_event: PhantomData::<fn(N)>,
+            };
+            extensions.insert(fmt_fields);
+        }
+    }
+
+    fn on_event(&self, event: &Event<'_>, ctx: Context<'_, S>) {
+        thread_local! {
+            static BUF: RefCell<String> = RefCell::new(String::new());
+        }
+
+        BUF.with(|buf| {
+            let borrow = buf.try_borrow_mut();
+            let mut a;
+            let mut b;
+            let mut buf = match borrow {
+                Ok(buf) => {
+                    a = buf;
+                    &mut *a
+                }
+                _ => {
+                    b = String::new();
+                    &mut b
+                }
+            };
+
+            let ctx = self.make_ctx(ctx);
+            if self.fmt_event.format_event(&ctx, &mut buf, event).is_ok() {
+                let mut writer = self.make_writer.make_writer();
+                let _ = io::Write::write_all(&mut writer, buf.as_bytes());
+            }
+
+            buf.clear();
+        });
+    }
+
+    unsafe fn downcast_raw(&self, id: TypeId) -> Option<*const ()> {
+        match () {
```

Comment:
I feel like a comment is warranted here. I have no idea why this code does what it does.

Reply:
this one isn't really @davidbarsky's fault either :) i think this method was copied directly from the existing subscriber code. i can add a comment!

Reply:
added a comment in cf796aebd27185eacac93c983bfc789617ad788f